spring-cloud-app-broker-core/src/test/java/org/springframework/cloud/appbroker/extensions/targets/TargetServiceTest.java

```diff
@@ -85,17 +91,64 @@ void shouldAddPropertiesToAllBackingApps() {
+	void shouldAddShortNameAsServiceInstanceGuidSuffix() {
+		final String appName = "app-name";
+		final String serviceInstanceId = UUID.randomUUID().toString();
+		final TargetSpec targetSpec = TargetSpec.builder().name("ServiceInstanceGuidSuffix").build();
+		final BackingApplication backingApplication = BackingApplication.builder().name(appName).build();
+
+		assertUsingAvailableLength(serviceInstanceId, appName);
+
+		StepVerifier.create(targetService.addToBackingApplications(singletonList(backingApplication), targetSpec,
+				serviceInstanceId))
+				.assertNext(backingApplications -> assertThat(backingApplications.get(0).getName())
+						.isEqualTo(assembleNameWithGuidSuffix(appName, serviceInstanceId)))
+				.verifyComplete();
+	}
 
-		List<BackingApplication> backingApplications =
-			targetService
-				.addToBackingApplications(singletonList(backingApplication), targetSpec, "guid")
-				.block();
+	@Test
+	void shouldAddLongNameAsServiceInstanceGuidSuffix() {
+		final String appName = "this-is-a-much-longer-app-name-that-will-require-truncation";
+		final String serviceInstanceId = UUID.randomUUID().toString();
+		final TargetSpec targetSpec = TargetSpec.builder().name("ServiceInstanceGuidSuffix").build();
+		final BackingApplication backingApplication = BackingApplication.builder().name(appName).build();
+
+		final String truncatedAppName = appName.substring(0, calculateAvailableLength(serviceInstanceId));
+		assertUsingAvailableLength(serviceInstanceId, truncatedAppName);
+
+		StepVerifier.create(targetService.addToBackingApplications(singletonList(backingApplication), targetSpec,
+				serviceInstanceId))
+				.assertNext(backingApplications -> assertThat(backingApplications.get(0).getName())
+						.isEqualTo(assembleNameWithGuidSuffix(truncatedAppName, serviceInstanceId)))
+				.verifyComplete();
+	}
+
+	@Test
+	void shouldAddMaxLengthNameAsServiceInstanceGuidSuffix() {
+		final String appName = "maxlengthname";
+		final String serviceInstanceId = UUID.randomUUID().toString();
+		final TargetSpec targetSpec = TargetSpec.builder().name("ServiceInstanceGuidSuffix").build();
+		final BackingApplication backingApplication = BackingApplication.builder().name(appName).build();
+
+		assertUsingAvailableLength(serviceInstanceId, appName);
+
+		StepVerifier.create(targetService.addToBackingApplications(singletonList(backingApplication), targetSpec,
+				serviceInstanceId))
+				.assertNext(backingApplications -> assertThat(backingApplications.get(0).getName())
+						.isEqualTo(assembleNameWithGuidSuffix(appName, serviceInstanceId)))
+				.verifyComplete();
+	}
+
+	private String assembleNameWithGuidSuffix(String appName, String serviceInstanceId) {
+		return appName + CONCAT_DELIM + serviceInstanceId;
+	}
+
+	private int calculateAvailableLength(String serviceInstanceId) {
+		return CF_SERVICE_NAME_MAX_LENGTH - CONCAT_DELIM.length() - serviceInstanceId.length();
+	}
```

Comment:
This essentially the same logic as in the production code. I personally tend to avoid using code logic to write test, so that we can validate the behavior in two different styles. The exact value don't actually matter, we just want to make sure the name ends up being unique and is less than 50 characters. So maybe we only need to make sure the name includes service instance id, has length of 50, and, optionally, starts with part of the service name. Wdyt?

Reply:
Yep, I see your point. Thanks! We do need to validate that case where the resulting name is less than 50 chars as well. While I don't like duplicating code, I do consider the goal of the test to be to validate the expected results. Even if the code to do that is basically the same now, the new name still needs to be determined somehow for comparison. But I'll reduce that duplication and update the PR.

Reply:
Right, it all depends on how we define the goal of this test. 
- If the behavior we are trying to describe here is exactly how we are truncating and concatenate names, then it will be hard to avoid duplicating the logic. 
- If the behavior is just keep it unique and less than 50 chars, then whether it's a `-` in between or how we truncate are not a behavioral requirements that needs to be tested. To achieve this goal, where the resulting name is less than 50 chars, we could assert that it contains the service instance id and the full service name.

Reply:
ok, redesigned the tests. thanks.